File: backend/secure_research/cache.py
from __future__ import annotations

import json
import time
from pathlib import Path

from .models import NormalizedPage
from .security import resolve_within_root


CACHE_FILE = "notion-corpus.json"


def load_cached_pages(
    root: Path, ttl_seconds: int
) -> list[NormalizedPage] | None:
    path = resolve_within_root(root, CACHE_FILE)
    if not path.exists():
        return None
    try:
        payload = json.loads(path.read_text(encoding="utf-8"))
        age = time.time() - float(payload["fetched_at"])
        if age > ttl_seconds:
            return None
        return [
            NormalizedPage.model_validate(page)
            for page in payload.get("pages", [])
        ]
    except (KeyError, TypeError, ValueError, json.JSONDecodeError):
        return None


def save_cached_pages(root: Path, pages: list[NormalizedPage]) -> None:
    path = resolve_within_root(root, CACHE_FILE)
    temporary = resolve_within_root(root, f"{CACHE_FILE}.tmp")
    payload = {
        "fetched_at": time.time(),
        "pages": [page.model_dump(mode="json") for page in pages],
    }
    temporary.write_text(
        json.dumps(payload, separators=(",", ":")), encoding="utf-8"
    )
    temporary.replace(path)

File: backend/server/src/cache.ts
import { rename } from "node:fs/promises";
import { resolveWithinRoot } from "./security/roots.ts";
import type { NormalizedPage } from "./types.ts";

const CACHE_FILE = "notion-corpus.json";

type CachePayload = {
  fetchedAt: number;
  pages: NormalizedPage[];
};

export async function loadCachedPages(
  root: string,
  ttlSeconds: number,
): Promise<NormalizedPage[] | null> {
  try {
    const payload = (await Bun.file(resolveWithinRoot(root, CACHE_FILE)).json()) as
      | CachePayload
      | undefined;
    if (!payload || Date.now() / 1000 - payload.fetchedAt > ttlSeconds) return null;
    return Array.isArray(payload.pages) ? payload.pages : null;
  } catch {
    return null;
  }
}

export async function saveCachedPages(
  root: string,
  pages: NormalizedPage[],
): Promise<void> {
  const path = resolveWithinRoot(root, CACHE_FILE);
  const temporary = resolveWithinRoot(root, `${CACHE_FILE}.tmp`);
  await Bun.write(
    temporary,
    JSON.stringify({ fetchedAt: Date.now() / 1000, pages } satisfies CachePayload),
  );
  await rename(temporary, path);
}
